### DISCUSSION TITLE
Best place to put initialization code.

### DISCUSSION THREAD

**Discussion by prabhavmehra** [dotnet/maui#20353] [Q&A] [answered]
Where is the best place to put initialization code to run it only once.

Currently we had it on the `OnLoaded` event but this actually had a bug in .net 7 where it only triggered once and this was fixed in .net 8. Similar issue, https://github.com/dotnet/maui/issues/16697.

Now with .net8 what is the correct and best place to initialize our code which only runs the first time irrespective of navigating to or away from the page

**Comment by Xyncgas** [comment]
MauiProgram.cs

**Comment by tschbc** [comment, marked as answer]
I remember reading someone's comment somewhere in this repo that you can use the Handler lifecycle for one-time "setup/teardown" methods. I don't remember where I saw it though.

This is what I use:

Create an extension so it's kept clean:

```c#
public static class HandlerChangingEventArgsExtension
{
    public static bool AttachingToHandler(this HandlerChangingEventArgs args)
    {
        return args.NewHandler != null && args.OldHandler == null;
    }

    public static bool DetachingFromHandler(this HandlerChangingEventArgs args)
    {
        return args.NewHandler == null && args.OldHandler != null;
    }
}
```

Create some base class for the Pages you use and override `OnHandlerChanging`:

```c#
protected override void OnHandlerChanging(HandlerChangingEventArgs args)
{
    base.OnHandlerChanging(args);

    if (args.AttachingToHandler())
        Creating();
    else if (args.DetachingFromHandler())
        Destroying();
}

protected virtual void Creating() { }

protected virtual void Destroying() { }
```

You can do whatever works for you with the Creating/Destroying methods.
